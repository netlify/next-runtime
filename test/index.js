const path = require('path')
const process = require('process')
const nextOnNetlify = require('next-on-netlify')
const pathExists = require('path-exists')
const { dir: getTmpDir } = require('tmp-promise')
const cpy = require('cpy')

const plugin = require('..')

const FIXTURES_DIR = `${__dirname}/fixtures`
const SAMPLE_PROJECT_DIR = `${__dirname}/sample`

const utils = {
  run: {
    command() {},
  },
  build: {
    failBuild(message) {
      throw new Error(message)
    },
  },
}

// Temporary switch cwd
const changeCwd = function (cwd) {
  const originalCwd = process.cwd()
  process.chdir(cwd)
  return process.chdir.bind(process, originalCwd)
}

// Move .next from sample project to current directory
const moveNextDist = async function () {
  await cpy('.next/**', process.cwd(), { cwd: SAMPLE_PROJECT_DIR, parents: true, overwrite: false, dot: true })
}

// Copy fixture files to the current directory
const useFixture = async function (fixtureName) {
  const fixtureDir = `${FIXTURES_DIR}/${fixtureName}`
  await cpy('**', process.cwd(), { cwd: fixtureDir, parents: true, overwrite: false, dot: true })
}

// In each test, we change cwd to a temporary directory.
// This allows us not to have to mock filesystem operations.
beforeEach(async () => {
  const { path, cleanup } = await getTmpDir({ unsafeCleanup: true })
  const restoreCwd = changeCwd(path)
  Object.assign(this, { cleanup, restoreCwd })
})

afterEach(async () => {
  jest.clearAllMocks()
  jest.resetAllMocks()

  // Cleans up the temporary directory from `getTmpDir()` and do not make it
  // the current directory anymore
  this.restoreCwd()
  await this.cleanup()
})

const DUMMY_PACKAGE_JSON = { name: 'dummy', version: '1.0.0' }

describe('preBuild()', () => {
  test('fail build if the app has static html export in npm script', async () => {
    await expect(
      plugin.onPreBuild({
        netlifyConfig: {},
        packageJson: { ...DUMMY_PACKAGE_JSON, scripts: { build: 'next export' } },
        utils,
        constants: { FUNCTIONS_SRC: 'out_functions' },
      }),
    ).rejects.toThrow('** Static HTML export next.js projects do not require this plugin **')
  })

  test('fail build if the app has static html export in toml/ntl config', async () => {
    await expect(
      plugin.onPreBuild({
        netlifyConfig: { build: { command: 'next build && next export' } },
        packageJson: DUMMY_PACKAGE_JSON,
        utils,
        constants: { FUNCTIONS_SRC: 'out_functions' },
      }),
    ).rejects.toThrow('** Static HTML export next.js projects do not require this plugin **')
  })

  test('fail build if app has next-on-netlify installed', async () => {
    const packageJson = {
      dependencies: { 'next-on-netlify': '123' },
    }
    await expect(
      plugin.onPreBuild({
        netlifyConfig: {},
        packageJson,
        utils,
      }),
    ).rejects.toThrow(
      `This plugin does not support sites that manually use next-on-netlify. Uninstall next-on-netlify as a dependency to resolve.`,
    )
  })

  test('fail build if app has next-on-netlify postbuild script', async () => {
    const packageJson = {
      scripts: { postbuild: 'next-on-netlify' },
    }
    await expect(
      plugin.onPreBuild({
        netlifyConfig: {},
        packageJson,
        utils,
      }),
    ).rejects.toThrow(
      `This plugin does not support sites that manually use next-on-netlify. Uninstall next-on-netlify as a dependency to resolve.`,
    )
  })

  test('fail build if the app has no package.json', async () => {
    await expect(
      plugin.onPreBuild({
        netlifyConfig: {},
        packageJson: {},
        utils,
        constants: { FUNCTIONS_SRC: 'out_functions' },
      }),
    ).rejects.toThrow(`Could not find a package.json for this project`)
  })

  test('create next.config.js with correct target if file does not exist', async () => {
    await plugin.onPreBuild({
      netlifyConfig: {},
      packageJson: DUMMY_PACKAGE_JSON,
      utils,
      constants: { FUNCTIONS_SRC: 'out_functions' },
    })

    expect(await pathExists('next.config.js')).toBeTruthy()
  })

  test.each(['invalid_next_config', 'deep_invalid_next_config'])(
    `fail build if the app's next config has an invalid target`,
    async (fixtureName) => {
      await useFixture(fixtureName)
      await expect(
        plugin.onPreBuild({
          netlifyConfig: {},
          packageJson: DUMMY_PACKAGE_JSON,
          utils,
          constants: { FUNCTIONS_SRC: 'out_functions' },
        }),
      ).rejects.toThrow(`next.config.js must be one of: serverless, experimental-serverless-trace`)
    },
  )
})

describe('onBuild()', () => {
  test('copy files to the publish directory', async () => {
    await useFixture('publish_copy_files')
    await moveNextDist()
    const PUBLISH_DIR = 'publish'
    await plugin.onBuild({
      constants: {
        PUBLISH_DIR,
        FUNCTIONS_SRC: 'functions',
      },
    })

    expect(await pathExists(`${PUBLISH_DIR}/_redirects`)).toBeTruthy()
    expect(await pathExists(`${PUBLISH_DIR}/index.html`)).toBeTruthy()
  })

  test.each([
    { FUNCTIONS_SRC: 'functions', resolvedFunctions: 'functions' },
    { FUNCTIONS_SRC: undefined, resolvedFunctions: 'netlify-automatic-functions' },
  ])('copy files to the functions directory', async ({ FUNCTIONS_SRC, resolvedFunctions }) => {
    await useFixture('functions_copy_files')
    await moveNextDist()
    await plugin.onBuild({
      constants: {
        FUNCTIONS_SRC,
        PUBLISH_DIR: '.',
      },
    })

    expect(await pathExists(`${resolvedFunctions}/next_api_test/next_api_test.js`)).toBeTruthy()
  })
})
