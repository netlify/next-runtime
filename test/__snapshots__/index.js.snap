// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onBuild() generates a file referencing all page sources 1`] = `
"// This file is purely to allow nft to know about these pages. It should be temporary.
exports.resolvePages = () => {
    try {
        require.resolve('../../../.next/server/pages/_app.js')
        require.resolve('../../../.next/server/pages/_document.js')
        require.resolve('../../../.next/server/pages/_error.js')
        require.resolve('../../../.next/server/pages/api/enterPreview.js')
        require.resolve('../../../.next/server/pages/api/exitPreview.js')
        require.resolve('../../../.next/server/pages/api/hello-background.js')
        require.resolve('../../../.next/server/pages/api/hello.js')
        require.resolve('../../../.next/server/pages/api/shows/[...params].js')
        require.resolve('../../../.next/server/pages/api/shows/[id].js')
        require.resolve('../../../.next/server/pages/deep/import.js')
        require.resolve('../../../.next/server/pages/getServerSideProps/[id].js')
        require.resolve('../../../.next/server/pages/getServerSideProps/all/[[...slug]].js')
        require.resolve('../../../.next/server/pages/getServerSideProps/static.js')
        require.resolve('../../../.next/server/pages/getStaticProps/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/env.js')
        require.resolve('../../../.next/server/pages/getStaticProps/static.js')
        require.resolve('../../../.next/server/pages/getStaticProps/with-revalidate.js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallback/[...slug].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallback/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallbackBlocking/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withRevalidate/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withRevalidate/withFallback/[id].js')
        require.resolve('../../../.next/server/pages/index.js')
        require.resolve('../../../.next/server/pages/middle/_middleware.js')
        require.resolve('../../../.next/server/pages/previewTest.js')
        require.resolve('../../../.next/server/pages/shows/[...params].js')
        require.resolve('../../../.next/server/pages/shows/[id].js')
    } catch {}
}"
`;

exports[`onBuild() generates a file referencing all page sources 2`] = `
"// This file is purely to allow nft to know about these pages. It should be temporary.
exports.resolvePages = () => {
    try {
        require.resolve('../../../.next/server/pages/_app.js')
        require.resolve('../../../.next/server/pages/_document.js')
        require.resolve('../../../.next/server/pages/_error.js')
        require.resolve('../../../.next/server/pages/api/enterPreview.js')
        require.resolve('../../../.next/server/pages/api/exitPreview.js')
        require.resolve('../../../.next/server/pages/api/hello-background.js')
        require.resolve('../../../.next/server/pages/api/hello.js')
        require.resolve('../../../.next/server/pages/api/shows/[...params].js')
        require.resolve('../../../.next/server/pages/api/shows/[id].js')
        require.resolve('../../../.next/server/pages/deep/import.js')
        require.resolve('../../../.next/server/pages/getServerSideProps/[id].js')
        require.resolve('../../../.next/server/pages/getServerSideProps/all/[[...slug]].js')
        require.resolve('../../../.next/server/pages/getServerSideProps/static.js')
        require.resolve('../../../.next/server/pages/getStaticProps/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/env.js')
        require.resolve('../../../.next/server/pages/getStaticProps/static.js')
        require.resolve('../../../.next/server/pages/getStaticProps/with-revalidate.js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallback/[...slug].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallback/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withFallbackBlocking/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withRevalidate/[id].js')
        require.resolve('../../../.next/server/pages/getStaticProps/withRevalidate/withFallback/[id].js')
        require.resolve('../../../.next/server/pages/index.js')
        require.resolve('../../../.next/server/pages/middle/_middleware.js')
        require.resolve('../../../.next/server/pages/previewTest.js')
        require.resolve('../../../.next/server/pages/shows/[...params].js')
        require.resolve('../../../.next/server/pages/shows/[id].js')
    } catch {}
}"
`;

exports[`onBuild() generates static files manifest 1`] = `Array []`;

exports[`onBuild() writes correct redirects to netlifyConfig 1`] = `
Array [
  Object {
    "from": "/_ipx/*",
    "status": 200,
    "to": "/.netlify/builders/_ipx",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/en/getStaticProps/withRevalidate/1.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/en/getStaticProps/withRevalidate/2.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/en/getStaticProps/withRevalidate/withFallback/1.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/en/getStaticProps/withRevalidate/withFallback/2.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/getStaticProps/with-revalidate.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/getStaticProps/with-revalidate.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/_next/data/build-id/getStaticProps/with-revalidate.json",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/_next/data/build-id/getStaticProps/withFallback/:id.json",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/_next/data/build-id/getStaticProps/withFallback/:slug/*",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/_next/data/build-id/getStaticProps/withFallbackBlocking/:id.json",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/_next/data/build-id/getStaticProps/withRevalidate/withFallback/:id.json",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/_next/image/*",
    "query": Object {
      "q": ":quality",
      "url": ":url",
      "w": ":width",
    },
    "status": 301,
    "to": "/_ipx/w_:width,q_:quality/:url",
  },
  Object {
    "from": "/_next/static/*",
    "status": 200,
    "to": "/static/:splat",
  },
  Object {
    "from": "/:locale/_next/static/*",
    "status": 200,
    "to": "/static/:splat",
  },
  Object {
    "conditions": Object {
      "Cookie": Array [
        "__prerender_bypass",
        "__next_preview_data",
      ],
    },
    "force": true,
    "from": "/*",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/*",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/api",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/api/*",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/BUILD_ID",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/build-manifest.json",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/cache/*",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/es/getStaticProps/with-revalidate",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/fr/getStaticProps/with-revalidate",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/getStaticProps/with-revalidate",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/getStaticProps/withFallback/:id",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/getStaticProps/withFallback/:slug/*",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "from": "/getStaticProps/withFallbackBlocking/:id",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "force": true,
    "from": "/getStaticProps/withRevalidate/1",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/getStaticProps/withRevalidate/2",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "from": "/getStaticProps/withRevalidate/withFallback/:id",
    "status": 200,
    "to": "/.netlify/builders/___netlify-odb-handler",
  },
  Object {
    "force": true,
    "from": "/getStaticProps/withRevalidate/withFallback/1",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/getStaticProps/withRevalidate/withFallback/2",
    "status": 200,
    "to": "/.netlify/functions/___netlify-handler",
  },
  Object {
    "force": true,
    "from": "/prerender-manifest.json",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/react-loadable-manifest.json",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/routes-manifest.json",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/server/*",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/serverless/*",
    "status": 404,
    "to": "/404.html",
  },
  Object {
    "force": true,
    "from": "/traces",
    "status": 404,
    "to": "/404.html",
  },
]
`;
